BUG: sparse: fix `idx_dtype` when building index arrays in many parts of sparse

dschult opened the pull request:
This PR adds `idx_dtype=get_index_dtype` handling in `random_array()` and `block_diag()`. It also rewrites the `kron` upcasting to int64 so that it uses `get_index_dtype` instead of handling it locally via `np.iinfo.max` checks.

Previously `random_array` always used `int64` for the index dtype. I have been wondering why so many test arrays become `int64` when I switch to `sparray`. They are `int32` for `spmatrix`. It turns out `random_array()` was providing `int64` index arrays to the `coo_array` constructor even for very small arrays. `random()` does this too, but they get downcast to `int32` during the final `spmatrix` construction for which `_get_index_dtype` checks input array contents and downcasts for small values. `sparray` in `_get_index_dtype` does not check array contents to determine idx_dtype. So with sparse arrays we have to downcast manually before construction.

I checked all functions in `_construct.py` and they all handle `idx_dtype` fine except `random_array()` and `block_diag()`. While I have not found `int64` from `block_diag`, it could have been using `int64` when `int32` is sufficient without anyone noticing, so I fixed it too. In addition to those two functions, the `kron()` function handles `idx_dtype` locally rather than using `get_index_dtype`, so I changed that to use the centralized code for uniformity and better long term maintenance. Lastly, I noticed a few places where `A.astype(dtype)` was used without the `copy=False` keyword just after `A` was created. So there is no chance a copy was required. I changed those to use `copy=False`.

I added some tests that downcasting to int32 occurs when appropriate. There are existing tests that upcasting occurs correctly (that behavior has always been fine). If there are better tests you can think of let me know. I think these are pretty good.

The incorrect behavior this fixes is subtle. The numerical values are still correct. But the resulting arrays cannot be used with `int32` libraries (such as SuperLU and some of csgraph's cython routines) without downcasting first. For a while I thought it was necessary to make the user manually downcast, but I realized that the `_construct.py` code **is** the user in this case and we can set up the incoming index arrays for `sparray` handling as `int32`. While adjusting that, I realized that `random_array` was implemented at the same time that the `sparray` constructors were changing from `check_contents=True` to `False` (summer 2023). They were parallel PRs and the new `random_array` behavior didn't account for the change in `check_contents` behavior.

I don't think this should require any changes to the doc_strings. We don't claim any particular behavior for the dtype of the indexing arrays.  It should make migration to sparray easier. I will put changes to the migration guide in a separate PR focusing on that process.

Fixes #20389

dschult committed:
fix idx_dtype treatment in random_array, block_diag and kron

perimosocordiae reviewed (dismissed):
Thanks for catching this subtle behavior, Dan.

Eventually, it would be nice to ensure that the index arrays are created with the minimal integer dtype at construction time, but downcasting before returning is a good first step.

dschult commented on scipy/sparse/_construct.py:
```diff
@@ -1233,7 +1238,7 @@ def data_sampler(size):
 
     # rng.choice uses int64 if first arg is an int
-    if tot_prod < np.iinfo(np.int64).max:
+    if tot_prod <= np.iinfo(np.int64).max:
```
This is unrelated to the others, but I noticed it because I was looking at use of `np.iinfo`.  `<` should be `<=`
Not gonna be noticed, especially because there is a workaround in the else clause. But in case someoen is **really** pushing the boundaries of int64. :)

dschult commented:
We could make an effort to change all building tools to choose the idx_type as early as possible. There are at least a few places where we might need to think a little about how to handle it. For example, random_array checks that `tot_prod` (the product of the shape) fits in `int64`. But it doesn't check if it fits in `int32`. That is the place where the `int64` arrays start for random_array. And we might need `int64` to select indices up to `tot_prod` even though `max(shape)` fits in `int32` because we unravel the random index integers after they. I guess we could use the slower backup index selection code (run when `tot_prod` > int64.max) and be even less likely to need int64. But there are speed trade-offs there.

I'm starting to circle around how to figure out where `int64` index arrays arise. Perhaps monkey patch `get_index_dtype` to raise when it returns `int64` and run the test suite.

We will also need more words in the migration guide about how to switch from relying on the `csr_matrix` constructor to downcast toward using `get_index_dtype` to get the right idx_dtype when the user is constructing the `coords` or `(indices, indptr)` info to give to `csr_array`. 

We might also choose to make a public tool like `_safe_downcast_index` from linalg (and now csgraph) that can check the array contents before casting for 32bit libraries like SuperLU, csgraph-cython, etc.  The possibility of csr_matrix running into this issue has always been there. But it seems to be showing up with sparray. That's what I'm trying to track down. Why are we running into the need for 32bit now.
:/

dschult committed:
handle idx_dtype of indices in linalg.spsolve, min_max, coo.reshape, random_array, block_diag, kron, cs.multiply

dschult committed:
linting

dschult commented on scipy/sparse/_construct.py:
```diff
@@ -938,7 +934,7 @@ def _block(blocks, format, dtype, return_spmatrix=False):
         A = _compressed_sparse_stack(blocks[:, 0], 0, return_spmatrix)
         if dtype is not None:
-            A = A.astype(dtype)
+            A = A.astype(dtype, copy=False)
```
not an idx_dtype issue, but good practice to avoid copy with `astype`

dschult commented on scipy/sparse/_construct.py:
```diff
@@ -951,7 +947,7 @@ def _block(blocks, format, dtype, return_spmatrix=False):
         A = _compressed_sparse_stack(blocks[0, :], 1, return_spmatrix)
         if dtype is not None:
-            A = A.astype(dtype)
+            A = A.astype(dtype, copy=False)
```
not an idx_dtype issue, but good practice to avoid copy with `astype`

dschult commented:
I've gone through the idx_dtype treatment in sparse (including csgraph and linalg) fairly thoroughly now.
I've identified functions that skipped handling idx_dtype during construction of indices. And I've improved the original changes proposed in this PR (by moving the handling idx_dtype earlier, e.g. inside `np.arange` or `np.concatenate`). 
I've moved the idx_dtype choice to construction time rather than casting where that is possible.

I've added these changes to this PR instead of a new PR because there were a fair number of improvements to the ones already here.  But it means I'm asking for a re-review. Sorry about that. Hopefully it is straightforward.

@perimosocordiae would you please re-review this PR?

dschult committed:
linting format strings

dschult committed:
fixes gh-20389: maintain idx_dtype across h/vstack. Also fixes block_diag

dschult commented:
This also now accommodates maintaining int64 idx_dtype across vstack/hstack/block_array and block_diag.
In particular for h/vstack it
Fixes #20389

perimosocordiae commented on scipy/sparse/_compressed.py:
```diff
@@ -994,12 +998,10 @@ def _prepare_indices(self, i, j):
             if idx < -bound:
-                raise IndexError('index (%d) out of range (< -%d)' %
-                                 (idx, bound))
+                raise IndexError('index ({idx}) out of range (< -{bound})')
```
```suggestion
                raise IndexError(f'index ({idx}) out of range (< -{bound})')
```

dschult committed:
fix f-string typos

dschult committed:
Merge branch 'main' into idx_dtype_construct

perimosocordiae commented:
Merged, thanks @dschult !
